# Individual Assignment #2  
Vincy Hu 

## Part A:  
a. Machine Learning Project: Generate frequent itemsets and association rules for a recommender system using
Association Rule Mining*. Feel free to choose appropriate levels of minimum support and minimum
confidence (e.g., 0.20 support and 30% confidence).

### Running Algorithm

In [1]:
# Import packages
import pandas as pd
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

In [8]:
# Set the column display to show all the columns 
pd.set_option('display.max_columns', None) 

In [49]:
# import data
trans = pd.read_excel('transactions_by_dept.xlsx')
trans

POS Txn                       Dept   ID  Sales U
0     16120100160021008773               0261:HOSIERY  250        2
1     16120100160021008773  0634:VITAMINS & HLTH AIDS  102        1
2     16120100160021008773          0879:PET SUPPLIES  158        2
3     16120100160021008773                 0973:CANDY  175        2
4     16120100160021008773               0982:SPIRITS  176        1
...                    ...                        ...  ...      ...
4534  16120100160162000843  0857:PC PERIPH/OFFICE ACC  155        1
4535  16120100160162000845   0395:MENS ATHLETIC SHOES   49        1
4536  16120100160162000845     0815:AUDIO ELECTRONICS  132        1
4537  16120100160162000846    0532:AMERICAN GREETINGS   72        1
4538  16120100160221001467  0066:VENDING/AMUSEMENT MA  242        1

[4539 rows x 4 columns]

In [5]:
trans.shape # learn about the dataset size: there are 4539 records of transaction 
            # (Note that one transaction will have multiple columns if there are multiple items in one transaction)

(4539, 4)

Before calculating the antecedents and consequents, we need to consolidate the items into one transsaction per row with each product.

In [9]:
basket = trans.groupby(['POS Txn','ID'])['Sales U'].sum().unstack().reset_index().fillna(0).set_index('POS Txn')
basket.head(5)

ID                    1    2    3    5    6    7    8    9    10   11   12   \
POS Txn                                                                       
16120100160021008773  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008774  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008775  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008776  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008777  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

ID                    13   14   17   18   19   21   22   23   24   26   27   \
POS Txn                                                                       
16120100160021008773  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008774  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008775  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008776  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008777  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

ID                    28   29   30   31   32   34   35   36   37   38   40   \
POS Txn                                                                       
16120100160021008773  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008774  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008775  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008776  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008777  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

ID                    41   42   44   45   46   48   49   50   51   52   53   \
POS Txn                                                                       
16120100160021008773  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008774  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008775  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008776  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008777  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

ID                    54   55   56   57   58   59   60   61   63   64   65   \
POS Txn                                                                       
16120100160021008773  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008774  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008775  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008776  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008777  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

ID                    70   71   72   73   74   75   76   77   78   79   80   \
POS Txn                                                                       
16120100160021008773  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008774  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008775  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008776  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008777  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

ID                    81   82   85   86   87   93   94   97   99   100  102  \
POS Txn                                                                       
16120100160021008773  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0   
16120100160021008774  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  5.0  0.0   
16120100160021008775  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008776  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
16120100160021008777  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

ID                    104  105  106  107  108  109  110  111  113  122  128  \
POS Txn                                    

In [14]:
# set all the value with >1 as a, set all other value as 0 in the basket dataframe
def set_quantity(x):
    if x>= 1:
        return 1
    if x<=0:
        return 0

basket_std = basket.applymap(set_quantity)
basket_std.head(5)

ID                    1    2    3    5    6    7    8    9    10   11   12   \
POS Txn                                                                       
16120100160021008773    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008774    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008775    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008776    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008777    0    0    0    0    0    0    0    0    0    0    0   

ID                    13   14   17   18   19   21   22   23   24   26   27   \
POS Txn                                                                       
16120100160021008773    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008774    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008775    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008776    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008777    0    0    0    0    0    0    0    0    0    0    0   

ID                    28   29   30   31   32   34   35   36   37   38   40   \
POS Txn                                                                       
16120100160021008773    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008774    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008775    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008776    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008777    0    0    0    0    0    0    0    0    0    0    0   

ID                    41   42   44   45   46   48   49   50   51   52   53   \
POS Txn                                                                       
16120100160021008773    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008774    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008775    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008776    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008777    0    0    0    0    0    0    0    0    0    0    0   

ID                    54   55   56   57   58   59   60   61   63   64   65   \
POS Txn                                                                       
16120100160021008773    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008774    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008775    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008776    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008777    0    0    0    0    0    0    0    0    0    0    0   

ID                    70   71   72   73   74   75   76   77   78   79   80   \
POS Txn                                                                       
16120100160021008773    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008774    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008775    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008776    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008777    0    0    0    0    0    0    0    0    0    0    0   

ID                    81   82   85   86   87   93   94   97   99   100  102  \
POS Txn                                                                       
16120100160021008773    0    0    0    0    0    0    0    0    0    0    1   
16120100160021008774    0    0    0    0    0    1    0    0    0    1    0   
16120100160021008775    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008776    0    0    0    0    0    0    0    0    0    0    0   
16120100160021008777    0    0    0    0    0    0    0    0    0    0    0   

ID                    104  105  106  107  108  109  110  111  113  122  128  \
POS Txn                                    

We will use **apriori** algorithm to build frequent itemsets.   
**Frequent itemset** is itemset that support is larger than the threshold set by marketing/sales team in the company.  
In **apriori** concept, frequent itemset are itemset that any combination of item in the itemset is also frequent itemset.The algorithm will iterate until excluding all non-frequent itemset in the itemsets.

In [46]:
# build frequent itemsets
frequent_itemsets = apriori(basket_std,min_support=0.03,use_colnames=True)

In [47]:
# display
frequent_itemsets

support    itemsets
0   0.049903        (72)
1   0.096899        (93)
2   0.063953        (99)
3   0.072674       (100)
4   0.031492       (108)
5   0.042151       (110)
6   0.076550       (143)
7   0.048934       (158)
8   0.047965       (169)
9   0.122578       (170)
10  0.133236       (175)
11  0.152132       (176)
12  0.092539       (177)
13  0.063953       (178)
14  0.089632       (179)
15  0.054264       (213)
16  0.031008   (100, 93)
17  0.032461  (170, 175)
18  0.037306  (176, 177)

To further filter the result from so many frequent itemset, we will use **lift** as the final filtering metrics.  
**Lift** measures how strong the association rule is. When the lift value is larger than 1, then it indicates that the association have more probability than randomness of the two items.

In [48]:
rules = association_rules(frequent_itemsets,metric = 'lift', min_threshold=1)
rules

antecedents consequents  antecedent support  consequent support   support  \
0       (100)        (93)            0.072674            0.096899  0.031008   
1        (93)       (100)            0.096899            0.072674  0.031008   
2       (170)       (175)            0.122578            0.133236  0.032461   
3       (175)       (170)            0.133236            0.122578  0.032461   
4       (176)       (177)            0.152132            0.092539  0.037306   
5       (177)       (176)            0.092539            0.152132  0.037306   

   confidence      lift  leverage  conviction  
0    0.426667  4.403200  0.023966    1.575176  
1    0.320000  4.403200  0.023966    1.363714  
2    0.264822  1.987610  0.016129    1.178985  
3    0.243636  1.987610  0.016129    1.160054  
4    0.245223  2.649948  0.023228    1.202290  
5    0.403141  2.649948  0.023228    1.420551

In [69]:
# select the unique item id and name list from transaction table
trans[trans['ID'] >= 177].sort_values('ID')

POS Txn                   Dept   ID  Sales U
1868  16120100160033009019              0983:WINE  177        1
3236  16120100160037008865              0983:WINE  177        1
3254  16120100160037008872              0983:WINE  177        4
729   16120100160023004246              0983:WINE  177        3
738   16120100160023004249              0983:WINE  177        1
...                    ...                    ...  ...      ...
4508  16120100160162000831        0860:NETWORKING  310       -1
4517  16120100160162000836        0860:NETWORKING  310        1
4436  16120100160162000788        0860:NETWORKING  310        1
2067  16120100160033009113        0860:NETWORKING  310        1
3741  16120100160077007448  0370:OPPORTUNITY BUYS  311       -2

[1035 rows x 4 columns]

93 - Health Aids   
100 - PERSONAL CARE  
170 - Beverage   
175 - Candy  
176 - Spitits  
177 - Wine   
 

### Business Advise

According to the association rule, I will recommend the business to :  
1) sell item 93 in association with item 100, put them nearby.  
2) sell item 170 in association with item 175, put them nearby.   
3) sell item 176 in association with item 177, put them nearby.   
Further, there are two pairs of association relationship are for beverage, the company can adjust the selling strategy to see whether they can promote beverage in different bundle.

# Part B